explorer/common/arena.h

```diff
@@ -61,35 +106,175 @@ class Arena {
-        : ArenaEntryTyped(std::forward<Args>(args)...) {}
+  // Factory metafunction for globally unique type IDs.
+  // &TypeId<T>::id == &TypeId<U>::id if and only if std::is_same_v<T,U>.
+  //
+  // Inspired by llvm::Any::TypeId.
+  template <typename T>
+  struct TypeId {
+    // This is only used for an address to compare; the value is unimportant.
+    static char id;
+  };
 
-    template <typename U, typename... Args>
-    explicit ArenaEntryTyped(WriteAddressTo<U> write_address, Args&&... args)
-        : ArenaEntryTyped(
-              (*write_address.target = &instance_, WriteAddressHelper{}),
-              std::forward<Args>(args)...) {}
+  // A canonicalization table maps a tuple of constructor argument values to
+  // a non-null pointer to a T object constructed with those arguments.
+  template <typename T, typename... Args>
+  using CanonicalizationTable =
+      std::unordered_map<std::tuple<ArgKeyType<Args>...>, Nonnull<const T*>,
+                         LlvmHasher>;
 
-    auto Instance() -> Nonnull<T*> { return Nonnull<T*>(&instance_); }
+  // Allocates an object in the arena. Unlike New, this will always allocate
+  // and construct a new object.
+  template <typename T, typename... Args>
+  auto UniqueNew(Args&&... args) -> Nonnull<T*>;
 
-   private:
-    T instance_;
-  };
+  // Returns a pointer to the canonical instance of T constructed from
+  // `args...`, or null if there is no such instance yet. Returns a mutable
+  // reference so that a null entry can be updated.
+  template <typename T, typename... Args>
+  auto CanonicalInstance(const Args&... args) -> const T*&;
 
   // Manages allocations in an arena for destruction at shutdown.
   std::vector<std::unique_ptr<ArenaEntry>> arena_;
   int64_t allocated_ = 0;
+
+  // Maps a CanonicalizationTable type to a unique instance of that type for
+  // this arena. For a key equal to &TypeId<T>::id for some T, the corresponding
+  // value contains a T*.
+  std::map<char*, std::any> canonical_tables_;
+};
+
+// ---------------------------------------
+// Implementation details only below here.
+// ---------------------------------------
+
+template <>
+struct ArgKey<std::nullopt_t> {
+  using type = struct NulloptProxy {
+    NulloptProxy(std::nullopt_t) {}
+    friend auto operator==(NulloptProxy, NulloptProxy) -> bool { return true; }
+    friend auto hash_value(NulloptProxy) -> llvm::hash_code {
+      return llvm::hash_combine();
+    }
+  };
+};
+
+template <typename T>
+struct ArgKey<std::vector<T>> {
+  using type = class VectorProxy {
+   public:
+    VectorProxy(std::vector<T> vec) : vec_(std::move(vec)) {}
+    friend auto operator==(const VectorProxy& lhs, const VectorProxy& rhs) {
+      return lhs.vec_ == rhs.vec_;
+    }
+    friend auto hash_value(const VectorProxy& v) {
+      return llvm::hash_combine(
+          llvm::hash_combine_range(v.vec_.begin(), v.vec_.end()),
+          v.vec_.size());
+    }
+
+   private:
+    std::vector<T> vec_;
+  };
 };
```

Comment:
I'm afraid so. This is for handling cases where a `vector<T>` is directly used as a constructor argument, so the machinery in `HashFromDecompose` doesn't have a chance to kick in. Even if it did, we wouldn't want it to, because here we need to make a deep copy of the `vector` argument so that we can retain it in the cache.